### PR TITLE
[FIX][16.0]account_payment_order: Limit available_partner_bank_ids to partner's one in OCA context

### LINKED ISSUE #1131: V16 account_banking_sepa_direct_debit
## Module
account_banking_sepa_direct_debit

## Describe the bug

When the SEPA file is generated, the debtor's IBAN is filled in with the creditor's IBAN.
Debtors' IBANs do not appear in the XML file

## To Reproduce
V16:

generate a SEPA DEBITOR xml file

**Expected behavior**
 the debtor's IBAN is filled in with the debtor's  IBAN

### LINKED ISSUE #1193: [16.0] On payment order Sepa Direct Debit, partner bank is not rirg
Same issue #1131 
reported by @cesarsc79 @10537

## Module
accoun_banking_sepa_direct_debit, but really the error is on account_payment_order module.

## Describe the bug
I can not replicate on runbot, but I'm not alone.

I create a payment order (sepa direct debit):
- invoice is ok
- customer is ok
- bank number is sepa and related with right bank, ok
- mandate is ok

When payment order is confirmed, the account.payment generated take the company account, not the partner account.

So the file generated is not right, because the company bank account is set DbtrAcct

## What's going on?

The function _prepare_account_payment_vals() is rigth, debuggin I can check that is the partner bank account

https://github.com/OCA/bank-payment/blob/55718af9a49b00598d06db2d6a045deb9cddd530/account_payment_order/models/account_payment_line.py#L184

But, partner_bank_account is computed stored:

```
    partner_bank_id = fields.Many2one('res.partner.bank', string="Recipient Bank Account",
        readonly=False, store=True, tracking=True,
        compute='_compute_partner_bank_id',
        domain="[('id', 'in', available_partner_bank_ids)]",
        check_company=True)
```

As sepa direct debit is incoming payment, the domain is set to company bank account.

https://github.com/odoo/odoo/blob/0ed63805e93a9e8ec059a496756dfe9b32e66a07/addons/account/models/account_payment.py#L403

```
    @api.depends('partner_id', 'company_id', 'payment_type', 'destination_journal_id', 'is_internal_transfer')
    def _compute_available_partner_bank_ids(self):
        for pay in self:
            if pay.payment_type == 'inbound':
                pay.available_partner_bank_ids = pay.journal_id.bank_account_id
            elif pay.is_internal_transfer:
                pay.available_partner_bank_ids = pay.destination_journal_id.bank_account_id
            else:
                pay.available_partner_bank_ids = pay.partner_id.bank_ids\
                        .filtered(lambda x: x.company_id.id in (False, pay.company_id.id))._origin

    @api.depends('available_partner_bank_ids', 'journal_id')
    def _compute_partner_bank_id(self):
        ''' The default partner_bank_id will be the first available on the partner. '''
        for pay in self:
            pay.partner_bank_id = pay.available_partner_bank_ids[:1]._origin
```

## How to fix it

I fix it overwriting _compute_partner_bank_id function to ensure that this field is not updated if I have a partner_bank_id set previously... this may be not the best way, but I'm not able to find why partner_bank_id is overwrite with company account.

```
class AccountPayment(models.Model):
    _inherit = "account.payment"
    
    @api.depends('available_partner_bank_ids', 'journal_id')
    def _compute_partner_bank_id(self):
        ''' The default partner_bank_id will be the first available on the partner. '''
        for pay in self.filtered(lambda p: not p.partner_bank_id):
            pay.partner_bank_id = pay.available_partner_bank_ids[:1]._origin
```

### PR DESCRIPTION
**Pull request by flotho**
FIX https://github.com/OCA/bank-payment/issues/1193
FIX #1131 
ping @angelmoya @alexis-via

### CODE REVIEW

**Comment by flotho** [comment]
how can I had a tag ready for review ?

**Comment by flotho** [comment]
hum... this depends method is not returning anything, yet pylint is insulting me.
How can I bypass this ?
Regards

**Comment by enriquemartin** [comment]
@flotho Could you rebase please?

**Review by enriquemartin — APPROVED** [review]

**Comment by flotho** [comment]
Hi @enriquemartin , I rebased.
Yet I still have a lint issue because I'm using a call to super.
Though the orginal method has no retrun so I'm wondering how to deal properly with this kind of inheritance ?
Regards

File: account_payment_order/models/account_payment.py [hunk 1]
```diff
@@ -87,3 +89,28 @@ def _prepare_move_line_default_vals(self, write_off_line_vals=None):
...
+        "company_id",
+        "payment_type",
+        "destination_journal_id",
+        "is_internal_transfer",
+    )
+    def _compute_available_partner_bank_ids(self):
+        original_payment_ids = self.filtered(
+            lambda pay: pay.payment_method_code not in OCA_PAYMENT_METHOD_CODES
+        )
+        super(
```
Comment (MiquelRForgeFlow) [inline account_payment_order/models/account_payment.py]:
```suggestion
        res = super(
```

File: account_payment_order/models/account_payment.py [hunk 2]
```diff
@@ -87,3 +89,28 @@ def _prepare_move_line_default_vals(self, write_off_line_vals=None):
...
+        oca_payment_ids = self - original_payment_ids
+        for pay in oca_payment_ids:
+            if pay.payment_method_code == "sepa_direct_debit":
+                pay.available_partner_bank_ids = pay.partner_id.bank_ids.filtered(
+                    lambda x: x.company_id.id in (False, pay.company_id.id)
+                )._origin
+            if pay.payment_method_code == "sepa_credit_transfer":
+                pay.available_partner_bank_ids = pay.partner_id.bank_ids.filtered(
+                    lambda x: x.company_id.id in (False, pay.company_id.id)
+                )._origin
```
Comment (MiquelRForgeFlow) [inline account_payment_order/models/account_payment.py:116]:
```python
        return res
```

Comment (flotho) [reply account_payment_order/models/account_payment.py:116]: thanks for the review @MiquelRForgeFlow 
I'm wondering what's the interest of returning something even if  the original method is not returning anything ?

Regards

**Comment by flotho** [comment]
Done !

**Comment by cubells** [comment]
I think this problem is solved with #1346

**Comment by pedrobaeza** [comment]
Yes, thanks.

**Pull request closed without merging**